Kernel.cc

```diff
@@ -1367,3 +1370,277 @@ short Kernel::findIndexNode( char * path , IndexNode& inode )
+
+	indexNodeNumber = FileSystem::ROOT_INDEX_NODE_NUMBER ;
+
+	token = NULL;
+	token = strtok(fullNewPath, "/");
+	memset(name, '\0', 512);
+	strcpy(name, ".");	//may be not needed.
+
+
+	while(1) {
+	    if(token != NULL) {
+		memset(name, '\0', 512);
+		strcpy(name, token);
+
+		// check to see if the current node is a directory
+
+		if((currIndexNode.getMode()&S_IFMT) != S_IFDIR) {
+		    // return (ENOTDIR) if a needed directory is not a directory
+		    process.errno = ENOTDIR ;
+		    return -1 ;
+		}
+
+		// get the next inode corresponding to the token
+		currIndexNode.copy(prevIndexNode);
+		
+		//Init CurIndexNode
+		emptyIndexNode.copy(currIndexNode);
+		indexNodeNumber = findNextIndexNode(openFileSystems, prevIndexNode, name, currIndexNode);
+		cout << "Name: " << name << " inode: " << indexNodeNumber << endl;
+	      }
+	    else {
+	      break;
+	    }
+
+	    token = strtok(NULL, "/");
+
+	    if(token != NULL) {
+		strcat(dirname, name);	
+		strcat(dirname, "/");
+	    }
+	}
+
+	fileDescriptor = NULL;
+	IndexNode sourceIndexNode;
+
+	// file does not exist.  We check to see if we can create it.
+	if ( indexNodeNumber < 0 ) {
+
+	    // Copy the source inode over
+	    fileSystem->readIndexNode(&sourceIndexNode, sourceIndexNodeNumber);
+
+	    // Increment nlinks
+	    int nlinks = sourceIndexNode.getNlink();
+	    nlinks++;
+	    sourceIndexNode.setNlink(nlinks);
+
+	    // Write inode
+	    fileSystem->writeIndexNode(&sourceIndexNode , sourceIndexNodeNumber);
+	    
+	    fileDescriptor = new FileDescriptor(fileSystem, sourceIndexNode , flags);
+	    // assign inode for the new file
+	    fileDescriptor->setIndexNodeNumber( sourceIndexNodeNumber ) ;
+
+	    // open the directory
+	    // instead of a name for the dir
+	    cout << "dirname: " << dirname << endl;
+	    int dir = open(dirname , O_RDWR);
+	    if( dir < 0 ) {
+		perror(PROGRAM_NAME);
+		cout << PROGRAM_NAME << ": unable to open directory for writing" << endl;
+		return -1;
+	    }
+
+	    // scan past the directory entries less than the current entry
+	    // and insert the new element immediately following
+	    int status = 0;
+	    DirectoryEntry newDirectoryEntry(sourceIndexNodeNumber, name);
+	    printf("Inode num: %d\n", sourceIndexNodeNumber);
+	    DirectoryEntry currentDirectoryEntry;
+
+	    while(true)
+	    {
+		// read an entry from the directory
+		status = readdir(dir, currentDirectoryEntry);
+
+		if(status < 0) {
+		    cout << PROGRAM_NAME << ": error reading directory in creat";
+		    exit( EXIT_FAILURE ) ;
+		}
+		else if( status == 0 ) {
+		    // if no entry read, write the new item at the current 
+		    // location and break
+		    writedir( dir , newDirectoryEntry) ;
+		    IndexNode note;
+		    
+		    static char funPath[1024];
+		    memset(fullPath, '\0', 1024);
+		    strcpy(funPath, "/away/jon/new.txt");
+		    short num = findIndexNode(funPath, note);
+		    //cout << "num is: " << num << endl;
```

Comment:
This is the line that is needed for some reason. It shouldn't be needed, but it's the only the filesystem seems to remember the inode number

Reply:
What is "/away/jon/new.txt"? It's not obvious to me what most of this is doing.

Reply:
Closed??? Seems harsh bro